src/flint/types/fmpq_mpoly.pyx

```diff
@@ -0,0 +1,982 @@
+        to fmpz) representing exponents, and corresponding coefficient values of fmpq.
+
+            >>> ctx = fmpq_mpoly_ctx.get_context(2,'lex','x,y')
+            >>> ctx.from_dict({(1,0):2, (1,1):3, (0,1):1})
+            3*x*y + 2*x + y
+        """
+        cdef:
+            fmpz_vec exp_vec
+            slong i, nvars = self.nvars()
+            fmpq_mpoly res
+
+        if not isinstance(d, dict):
+            raise ValueError("expected a dictionary")
+
+        res = create_fmpq_mpoly(self)
+
+        for i, (k, v) in enumerate(d.items()):
+            v = any_as_fmpq(v)
+            if v is NotImplemented:
+                raise TypeError(f"cannot coerce coefficient '{v}' to fmpq")
+            elif len(k) != nvars:
+                raise ValueError(f"expected {nvars} exponents, got {len(k)}")
+
+            exp_vec = fmpz_vec(k)
+
+            if v:
+                fmpq_mpoly_push_term_fmpq_ffmpz(res.val, (<fmpq>v).val, exp_vec.val, self.val)
+
+        fmpq_mpoly_sort_terms(res.val, self.val)
+        fmpq_mpoly_reduce(res.val, self.val)
+        return res
+
+
+cdef class fmpq_mpoly(flint_mpoly):
+    """
+    The *fmpq_mpoly* type represents sparse multivariate polynomials over
+    the integers.
+    """
+
+    def __cinit__(self):
+        self._init = False
+
+    def __dealloc__(self):
+        if self._init:
+            fmpq_mpoly_clear(self.val, self.ctx.val)
+            self._init = False
+
+    def __init__(self, val=0, ctx=None):
+        if typecheck(val, fmpq_mpoly):
+            if ctx is None or ctx == (<fmpq_mpoly>val).ctx:
+                init_fmpq_mpoly(self, (<fmpq_mpoly>val).ctx)
+                fmpq_mpoly_set(self.val, (<fmpq_mpoly>val).val, self.ctx.val)
+            else:
+                raise IncompatibleContextError(f"{ctx} is not {(<fmpq_mpoly>val).ctx}")
+        elif typecheck(val, fmpz_mpoly):
+            if ctx is None:
+                ctx = fmpq_mpoly_ctx.from_context((<fmpz_mpoly>val).ctx)
+            elif not typecheck(ctx, fmpz_mpoly_ctx):
+                raise TypeError(f"{ctx} is not a fmpz_mpoly_ctx")
+            init_fmpq_mpoly(self, ctx)
+            fmpz_mpoly_set(self.val.zpoly, (<fmpz_mpoly>val).val, (<fmpz_mpoly> val).ctx.val)
+            fmpq_one(self.val.content)
+            fmpq_mpoly_reduce(self.val, self.ctx.val)
+        elif isinstance(val, dict):
+            if ctx is None:
+                if len(val) == 0:
+                    raise ValueError("Need context for zero polynomial")
+                k = list(val.keys())[0]
+                if not isinstance(k, tuple):
+                    raise ValueError("Dict should be keyed with tuples of integers")
+                ctx = fmpq_mpoly_ctx.get_context(len(k))
+            x = ctx.from_dict(val)
+            # XXX: this copy is silly, have a ctx function that assigns an fmpz_mpoly_t
+            init_fmpq_mpoly(self, ctx)
+            fmpq_mpoly_set(self.val, (<fmpq_mpoly>x).val, self.ctx.val)
```

Comment:
I'm not sure about creating a context on the fly like this.

Reply:
I've remove that context creatation, and the creation in `fmpz_mpoly`. However there is another one here
```python
        elif typecheck(val, fmpz_mpoly):
            if ctx is None:
                ctx = fmpq_mpoly_ctx.from_context((<fmpz_mpoly>val).ctx)
            elif not typecheck(ctx, fmpq_mpoly_ctx):
                raise TypeError(f"{ctx} is not a fmpq_mpoly_ctx")
            elif ctx.nvars != val.context().nvars():
                raise ValueError(
                    f"Provided context ({ctx}) and provided fmpz_mpoly ({val}) don't share the same number of variables"
                )
            init_fmpq_mpoly(self, ctx)
            fmpz_mpoly_set(self.val.zpoly, (<fmpz_mpoly>val).val, (<fmpz_mpoly> val).ctx.val)
            fmpq_one(self.val.content)
            fmpq_mpoly_reduce(self.val, self.ctx.val)
```
I've just added the `nvars` check but this is a little sketchy, we're allowing the user to provided a `fmpz_mpoly` and a `fmpq_mpoly_ctx` (which is really a `fmpz_mpoly_ctx`) and change the context under the polys feet. Ideally the we'd just check the `fmpq_mpoly_ctx` is just the same object as the `fmpz_mpoly_ctx` but our interface does not allow for that.